File: 1110-delete-nodes-and-return-forest/1110-delete-nodes-and-return-forest.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
    void dfs(TreeNode* &root, unordered_set<int> &s, vector<TreeNode*> &ans) {
        if(!root) return;
        
        dfs(root -> left, s, ans);
        dfs(root -> right, s, ans);
        
        if(s.find(root -> val) != s.end()) {
            if(root -> left) ans.push_back(root -> left);
            if(root -> right) ans.push_back(root -> right);
            
            root = NULL;
            delete root;
        }
    }
    
public:
    vector<TreeNode*> delNodes(TreeNode* root, vector<int>& to_delete) {
        unordered_set<int> s;
        s.insert(to_delete.begin(), to_delete.end());
        vector<TreeNode*> ans;
        
        dfs(root, s, ans);
        if(root) ans.push_back(root);
        
        return ans;
    }
};

File: 1110-delete-nodes-and-return-forest/1110-delete-nodes-and-return-forest.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    
    public boolean f(TreeNode root,Set<Integer> st, List<TreeNode> lt){ 
        if(root==null) return false;
        
        if(f(root.left,st,lt))
            root.left=null;
        
        if(f(root.right,st,lt))
            root.right=null;
        
        if(st.contains(root.val)){ 
            if(root.left!=null) lt.add(root.left);
            if(root.right!=null) lt.add(root.right);
            return true;
        }
        return false;
    }
    public List<TreeNode> delNodes(TreeNode root, int[] nums) {
        List<TreeNode> lt=new LinkedList<>();
        Set<Integer> st=new HashSet<>();
        for(int e:nums) st.add(e);
        boolean vis=f(root,st,lt);
        if(!vis) lt.add(root);
        return lt;
    }
}